### DISCUSSION TITLE
Give easy access to the scoped class name

### DISCUSSION THREAD

**Discussion by IcyFoxe** [withastro/roadmap#570] [Proposal]
I love using Astro components, but my biggest gripe is having no access to the unique **generated class name**, like `astro-CYLJZ4DT`.
It would be really helpful if it was possible to easily access this generated ID, for example like:

```js
const scopeId = Astro.id; // CYLJZ4DT
```

That way it would be possible to target only the elements within the scope more easily:

```js
const componentContainer = document.querySelector(`.container.astro-${scopeId}`);
```

**Comment by delucis** [comment]
Definitely an idea that has come up. You can’t use the `Astro` global in `<script>` tags currently, so this proposal wouldn’t be a perfect fix. It’s actually trickier than it looks because a bundled script runs once, but what you’re describing is most likely code you want to run once _per component instance_.

I’ll shout out that custom elements (aka “web components”) are a good way to solve this currently: https://docs.astro.build/en/guides/client-side-scripts/#web-components-with-custom-elements

**Comment by IcyFoxe** [comment]
Ah no, now I see. Every component, even if used multiple times, has the same class, so even if I would have access to that class id from within, it wouldn't make any difference. On the other hand, I did try the custom elements approach before, and it has some advantages, especially for element targeting, but for CSS, I'm not sure it makes any difference.

I'll describe you what I was trying to achieve, and then maybe you could help me suggest some approach, how to solve it.

I imagined something like an image component that accepts `maxWidth` property to set the max width of an image. However I wanted to make it responsive, so you would instead provide an object, where key would represent the breakpoint and value would be the width:
```html
<Image maxWidth={{ sm: '100%', md: '50%', lg: '300px' }}></Image>
```
That's why I wanted to create component-instance specific CSS using the unique class id, so that I could set the breakpoints based on the provided values.

But this turned out to be tougher to achieve than I though. If you have any suggestion how could this be solved, I would be happy to learn.

**Reply by delucis** [reply]
This is probably a good use case for [the `define:vars` directive](https://docs.astro.build/en/reference/directives-reference/#definevars):

```astro
---
const { maxWidth } = Astro.props;
const { sm, md, lg } = maxWidth;
---

<!-- placeholder for your actual image component -->
<div class="component" />

<style define:vars={{ sm, md, lg }}>
.component {
  max-width: var(--sm, 100%);
}

@media (min-width: 30em) {
  .component {
    max-width: var(--md, 50%);
  }
}

@media (min-width: 50em) {
  .component {
    max-width: var(--lg, 300px);
  }
}
</style>
```

**Comment by IcyFoxe** [comment]
Hmm interesting. I didn't know about fallback values within the CSS var function. Though does this mean that if default `maxWidth` value was `100%` (for all breakpoints), I would have to specify `100%` as a fallback value on each breakpoint? That feels a bit cumbersome..
Though I could use SCSS mixins or something for that.

By the way, why when you use `define:vars` on a `<script>` tag, TypeScript stops working?

**Reply by delucis** [reply]
Check out the caution box here for why TypeScript isn’t available in scripts with `define:vars`: https://docs.astro.build/en/reference/directives-reference/#definevars

**Comment by IcyFoxe** [comment]
After testing out the approach you suggested, it has a problem.
If I specify only for example a medium breakpoint:
```
<Image maxWidth={{ md: '400px' }}></Image>
```
and define CSS like you did, it will work only for that specific breakpoint and get overriden on larger breakpoints - in this case the `lg` breakpoints. That's no good.

**Reply by delucis** [reply]
You can write additional logic for your specific use case — my code was just an example.

If larger breakpoints should inherit smaller breakpoint values, you could do something like this for example:

```js
const { sm, md = sm, lg = md } = maxWidth;
```

This uses the value of `sm` for `md` if `md` is not set, and the value of `md` for `lg` if `lg` is not set.

Similarly if your default is always `100%`, you could also skip the fallback values in the `style` block and set that in the same place:

```js
const { sm = '100%', md = sm, lg = md } = maxWidth;
```

Like this you’re basically guaranteeing all your breakpoints have values, so can skip adding fallback values in the CSS.

**Comment by IcyFoxe** [comment]
Ah right, thank you for all your help. I do appreciate it!
